File: Exercises/ExercisesFortran/11Jan_Chapter07/4-BarrierCrossing/Source/integrate.f
      Subroutine Integrate(DeltaUtotal)
      Implicit None
 
      Include 'system.inc'
 
Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Integrate The Equations Of Motion For An Nve System  C
C     Velocity Verlet Integrator                           C
C                                                          C
C     DeltaUtotal Is The Energy Drift                             C
Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc

      Integer I
      Double Precision U,F,Consv,Utotal0,DeltaUtotal,DeltaUcounter
 
      Call Force(Xpos,U,F)

      DeltaUtotal = 0.0d0
      DeltaUcounter = 0.0d0
      Utotal0 = 0.0d0

Cccccccccccccccccccccccccccccccccccccccccccccc
C     Integrate The E.O.M. For Nstep Steps   C
Cccccccccccccccccccccccccccccccccccccccccccccc

      Do I=1,Nstep
         Xpos = Xpos             + Vpos*Tstep      + 0.5d0*F*Tstep*Tstep
         Vpos = Vpos             + 0.5d0*Tstep*F

         Call Force(Xpos,U,F)

         Vpos  = Vpos            + 0.5d0*Tstep*F
         Consv = 0.5d0*Vpos*Vpos + U

Ccccccccccccccccccccccccccccccccccccccccccccccc
C     Used For Calculating The Energy Drift   C
C     Not Very Important Really...            C
Ccccccccccccccccccccccccccccccccccccccccccccccc

         If(I.Eq.1) Then
            Utotal0 = Consv
         Else
            If(.Not.(I.Gt.5.And.
     &               ((Xpos.Lt.-1.0d0.And.Vpos.Lt.0.0d0).Or.
     &                (Xpos.Gt.1.0d0.And.Vpos.Gt.0.0d0)))) Then

               DeltaUtotal = DeltaUtotal + Dabs((Consv-Utotal0)/Utotal0)
               DeltaUcounter = DeltaUcounter + 1.0d0
            Endif
         Endif

         Call Sample(2,I)

Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Check If The Particle Passed The Boundary Or If It Can Never Recross The   C
C     Boundary Again...                                                          C
Cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc

      Enddo

      If(DeltaUcounter.Gt.0.5d0) Then
         DeltaUtotal = DeltaUtotal/DeltaUcounter
      Else
         DeltaUtotal = 0.0d0
      Endif

      Return
      End

File: Exercises/ExercisesC/11Jan_Chapter07/4-BarrierCrossing/Source/integrate.c
#include <stdlib.h> 
#include <stdio.h> 
#include <math.h> 
#include "system.h"
 
// integrate the equations of motion for an nve system
// velocity verlet integrator
// DeltaUtotal is the energy drift
void Integrate(double *DeltaUdrift)
{
  int i;
  double U,F,Consv,Utotal0,DeltaUtotal,DeltaUcounter;
 
  Force(Xpos,&U,&F);

  DeltaUtotal=0.0;
  DeltaUcounter=0.0;
  Utotal0=0.0;

  // integrate the E.O.M. for Nstep steps
  for(i=0;i<NumberOfSteps;i++)
  {
    Xpos+=Vpos*Tstep+0.5*F*Tstep*Tstep;
    Vpos+=0.5*Tstep*F;

    Force(Xpos,&U,&F);

    Vpos+=0.5*Tstep*F;
    Consv=0.5*Vpos*Vpos+U;

    // used for calculating the energy drift

    if(i==0)
      Utotal0=Consv;
    else
    {
      if(!(i>5&&((Xpos<-1.0&&Vpos<0.0)||(Xpos>1.0&&Vpos>0.0))))
      {
        DeltaUtotal+=fabs((Consv-Utotal0)/Utotal0);
        DeltaUcounter+=1.0;
      }
    }

    Sample(2,i);
  }

  if(DeltaUcounter>0.5)
    *DeltaUdrift=DeltaUtotal/DeltaUcounter;
  else
    *DeltaUdrift=0.0;
}
